File: leetcode/P50.java
class Solution {
    public double myPow(double x, int n) {
        if (n == 0) return 1;
        if (n == 1) return x;
        if (n == -1) return 1.0/x;
        
        double result = myPow(x, n/2);
        if (n%2 == 0) {
            return result * result;
        } else {
            if (n > 0)
                return result * result * x;
            else
                return result * result * 1.0/x;
        }
    }
}

File: leetcode/P50.py
def powUtil(x, n):
    if n == 0 or n == 1:
        return 1 if n is 0 else x
    if n % 2 == 0:
        val = powUtil(x, n//2)
        return val * val
    else:
        val = powUtil(x, n//2)
        return x * val * val


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 1:
            return 1
        if n < 0:
            return 1.00000/(powUtil(x, -n))
        else:
            return powUtil(x, n)
